File: practice/leetcode/blind75/200_number_of_islands.py
from typing import List
class Solution:
    def numOfIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        answer = 0
        rows, cols = len(grid), len(grid[0])

        def dfs(r, c):
            grid[r][c] = '2'

            for d_r, d_c in (0, 1), (0, -1), (1, 0), (-1, 0):
                new_r = r + d_r
                new_c = c + d_c
                if 0 <= new_r < rows and 0 <= new_c < cols and grid[new_r][new_c] == '1':
                    dfs(new_r, new_c)

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    dfs(r, c)
                    answer += 1

        return answer

File: practice/leetcode/blind75/golang/200_number_of_islands.go
func numOfIslands(grid [][]int) int {
	answer := 0
	for r := 0; r < len(grid); r++ {
		for c := 0; c < len(grid[r]); c++ {
			if grid[r][c] == 1 {
				dfs(grid, i, j)
				answer += 1
			}
		}
	}
}

func dfs(grid [][]int, r, c int) {
	if r < 0 || r >= len(grid) || c < 0 || c >= len(grid[r]) || grid[r][c] != 1 {
		return
	}
	dfs(grid, r+1, c)
	dfs(grid, r-1, c)
	dfs(grid, r, c+1)
	dfs(grid, r, c-1)
}
